compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -2829,32 +2829,68 @@ private BLangBlockStmt desugarForeachToWhile(BLangForeach foreach, BLangSimpleVa
                 getIteratorNextVariableDefinition(foreach, iteratorSymbol, resultSymbol);
 
         // Note - $result$ != ()
+        BLangType userDefineType;
+        if (foreach.errorType != null) {
+            userDefineType = getUserDefineTypeNode(BUnionType.create(null,
```

Comment:
We can just pass the member types to BUnionType.create(null, foreach.resultType, foreach.errorType)
let's change that when you add the test cases..

Reply:
Fixed in 1e740aab3c357fd933523708d6fa925996fcab0b